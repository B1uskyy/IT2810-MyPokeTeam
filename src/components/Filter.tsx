import React, { useEffect, useState } from 'react';
import '../styles/filter.css';

type FilterProps = {
  name: string;
  values: string[];
  onFilterChange: (selectedValue: string) => void;
};

export default function Filter({ name, values, onFilterChange }: FilterProps) {
  const [selectedValue, setSelectedValue] = useState('');

  const handleChange = (e: React.ChangeEvent<HTMLSelectElement>) => {
    const selected = e.target.value;
    setSelectedValue(selected);
    onFilterChange(selected);
    sessionStorage.setItem('filter', selected);
  };

  useEffect(() => {
    const savedFilter = sessionStorage.getItem('filter');
    if (savedFilter) {
      setSelectedValue(savedFilter);
    }
  }, [selectedValue]);
  return (
    <div className="pokemon-types">
      <label htmlFor={name.toLowerCase()}>Filter on type</label>
      <select id={name.toLowerCase()} value={selectedValue} onChange={handleChange} aria-label={`Select a ${name} `}>
        <option value="">All types</option>
        {values.map((value, index) => (
          <option key={index} value={value}>
            {value}
          </option>
        ))}
      </select>
    </div>
  );
}
