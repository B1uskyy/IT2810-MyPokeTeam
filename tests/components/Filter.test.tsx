import { describe, it, expect } from "vitest";

import { render, screen, fireEvent } from '@testing-library/react';
import Filter from '../../src/components/Filter';
import { vi } from 'vitest';
import { MemoryRouter, Route, Routes } from 'react-router-dom';

const mockOnFilterChange = vi.fn();

const createTestFilterData = () => ({
  name: 'Type',
  values: ['Fire', 'Water', 'Grass'],
});

let testFilterData: ReturnType<typeof createTestFilterData>;

beforeEach(() => {
  testFilterData = createTestFilterData();
  mockOnFilterChange.mockClear();
});

describe('Filter component', () => {
  it('renders with correct props ', () => {
    render(
      <MemoryRouter initialEntries={['/']}>
        <Routes>
          <Route
            path="/"
            element={<Filter name={testFilterData.name} values={testFilterData.values} onFilterChange={mockOnFilterChange} />}
          />
        </Routes>
      </MemoryRouter>,
    );

    expect(screen.getByLabelText(`Select a ${testFilterData.name}`)).toBeInTheDocument();
  });

  it('matches snapshot with correct props', () => {
    const { asFragment } = render(
      <MemoryRouter initialEntries={['/']}>
        <Routes>
          <Route
            path="/"
            element={<Filter name={testFilterData.name} values={testFilterData.values} onFilterChange={mockOnFilterChange} />}
          />
        </Routes>
      </MemoryRouter>,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders correct number of options', () => {
    render(
      <MemoryRouter initialEntries={['/']}>
        <Routes>
          <Route
            path="/"
            element={<Filter name={testFilterData.name} values={testFilterData.values} onFilterChange={mockOnFilterChange} />}
          />
        </Routes>
      </MemoryRouter>,
    );

    testFilterData.values.forEach((value) => {
      expect(screen.getByText(value)).toBeInTheDocument();
    });

    const options = screen.getAllByRole('option');
    expect(options).toHaveLength(testFilterData.values.length + 1);
  });

  it('calls onFilterChange when a value is selected', () => {
    render(
      <MemoryRouter initialEntries={['/']}>
        <Routes>
          <Route
            path="/"
            element={<Filter name={testFilterData.name} values={testFilterData.values} onFilterChange={mockOnFilterChange} />}
          />
        </Routes>
      </MemoryRouter>,
    );

    const selectElement = screen.getByLabelText(`Select a ${testFilterData.name}`) as HTMLSelectElement;

    fireEvent.change(selectElement, { target: { value: 'Fire' } });

    expect(mockOnFilterChange).toHaveBeenCalledWith('Fire');
    expect(selectElement.value).toBe('Fire');
  });

  it('persists filter value in sessionStorage', () => {
    render(
      <MemoryRouter initialEntries={['/']}>
        <Routes>
          <Route
            path="/"
            element={<Filter name={testFilterData.name} values={testFilterData.values} onFilterChange={mockOnFilterChange} />}
          />
        </Routes>
      </MemoryRouter>,
    );

    const selectElement = screen.getByLabelText(`Select a ${testFilterData.name}`);

    // Simulate selecting a value
    fireEvent.change(selectElement, { target: { value: 'Water' } });

    // Check if the selected value is saved to sessionStorage
    expect(sessionStorage.getItem('filter')).toBe('Water');
  });

  it('restores filter value from sessionStorage', () => {
    // Set a value in sessionStorage before rendering
    sessionStorage.setItem('filter', 'Grass');

    render(
      <MemoryRouter initialEntries={['/']}>
        <Routes>
          <Route
            path="/"
            element={<Filter name={testFilterData.name} values={testFilterData.values} onFilterChange={mockOnFilterChange} />}
          />
        </Routes>
      </MemoryRouter>,
    );

    const selectElement = screen.getByLabelText(`Select a ${testFilterData.name}`) as HTMLSelectElement;

    expect(selectElement.value).toBe('Grass');
  });
});