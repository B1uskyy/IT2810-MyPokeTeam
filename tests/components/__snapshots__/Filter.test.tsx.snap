// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Filter component > matches snapshot with correct props 1`] = `
<DocumentFragment>
  <div
    class="pokemon-types"
  >
    <label
      for="type"
    >
      Type
    </label>
    <select
      aria-label="Select a Type "
      id="type"
    >
      <option
        value=""
      >
        Select Type
      </option>
      <option
        value="Fire"
      >
        Fire
      </option>
      <option
        value="Water"
      >
        Water
      </option>
      <option
        value="Grass"
      >
        Grass
      </option>
    </select>
  </div>
</DocumentFragment>
`;
